ts/private/ts_project.bzl

```diff
@@ -215,81 +215,113 @@ This is an error because Bazel does not run actions unless their outputs are nee
-
+        typecheck_arguments = ctx.actions.args()
         if supports_workers:
-            arguments.add("--bazelValidationFile", typecheck_output.short_path)
+            typecheck_arguments.add("--bazelValidationFile", typecheck_output.short_path)
 
-        arguments.add("--noEmit")
+        typecheck_arguments.add("--noEmit")
+
+        ctx.actions.run(
+            executable = executable,
+            inputs = transitive_inputs_depset,
+            arguments = [arguments, typecheck_arguments],
+            outputs = [typecheck_output],
+            mnemonic = "TsProjectCheck",
+            execution_requirements = execution_requirements,
+            resource_set = resource_set(ctx.attr),
+            progress_message = "Type-checking TypeScript project %s [tsc -p %s]" % (
+                ctx.label,
+                tsconfig_path,
+            ),
+            env = {
+                "BAZEL_BINDIR": ctx.bin_dir.path,
+                "JS_BINARY__STDOUT_OUTPUT_FILE": typecheck_output.path,
+            },
+        )
     else:
         typecheck_outs.extend(output_types)
 
-    inputs_depset = depset()
-    if len(outputs) > 0:
-        transitive_inputs.append(_gather_types_from_js_infos(srcs_tsconfig_deps))
+    if use_tsc_for_js or use_tsc_for_dts:
+        tsc_emit_arguments = ctx.actions.args()
 
-        inputs_depset = depset(
-            copy_files_to_bin_actions(ctx, inputs),
-            transitive = transitive_inputs,
-        )
+        # Emitting without type-checking which will be done async by a separate action.
+        if ctx.attr.async_typecheck:
+            tsc_emit_arguments.add("--noCheck")
+            tsc_emit_arguments.add("--skipLibCheck")
+            tsc_emit_arguments.add("--noResolve")
 
-        if ctx.attr.transpile != 0 and not ctx.attr.emit_declaration_only and not ctx.attr.no_emit:
-            if ctx.attr.declaration:
-                verb = "Transpiling & type-checking"
-            else:
-                verb = "Transpiling"
-        else:
-            verb = "Type-checking"
+        # Emitting only declarations
+        if not use_tsc_for_js:
+            tsc_emit_arguments.add("--emitDeclarationOnly")
+
+        # Not emitting declarations
+        if not use_tsc_for_dts:
+            # TODO: why doesn't this work with workers?
+            if not supports_workers:
+                tsc_emit_arguments.add("--declaration", "false")
 
-        env = {
-            "BAZEL_BINDIR": ctx.bin_dir.path,
-        }
+        verb = "Transpiling" if ctx.attr.async_typecheck else "Transpiling & type-checking"
 
-        if stdout_file != "":
-            env["JS_BINARY__STDOUT_OUTPUT_FILE"] = stdout_file
+        inputs_depset = inputs if ctx.attr.async_typecheck else transitive_inputs_depset
 
         ctx.actions.run(
             executable = executable,
             inputs = inputs_depset,
-            arguments = [arguments],
+            arguments = [arguments, tsc_emit_arguments],
```

Comment:
`--emitDeclarationOnly` can appear twice since both `arguments` and `tsc_emit_arguments` can contain `--emitDeclarationOnly`

Reply:
I think you can ignore this. Seeing `--emitDeclarationOnly` twice is probably a side-effect of the other issue I commented on.

Reply:
I'll verify that case more when I add proper tests for this 👍

Reply:
I think this was fixed in #679